### ISSUE TITLE
Clean shutdown not possible due to lingering executors

### ISSUE THREAD

**Issue by bgooren** [skuzzle/guice-async-extension#3]
Thank you for an easy to use library!

I have one issue with it though: a clean shutdown is currently not possible as the default ```ScheduledExecutorService``` created by ```AsyncModule``` is never shut down.
Can you please provide a method to perform a clean shutdown? E.g. ```GuiceAsync.shutdown()```

I am currently doing this, but it's hackish:

```
Class< ? extends Annotation> defaultBindingAnnotation;
Injector injector = ...
try
        {
            defaultBindingAnnotation = (Class< ? extends Annotation>) Class.forName( "de.skuzzle.inject.async.internal.DefaultBinding" );
        }
        catch( ClassNotFoundException e )
        {
            throw new RuntimeException( "Cannot find DefaultBinding annotation!" );
        }

ExecutorService executorService = injector.getInstance( Key.get( ExecutorService.class, defaultBindingAnnotation ) );
        executorService.shutdownNow();

        ScheduledExecutorService scheduledExecutorService = injector
                .getInstance( Key.get( ScheduledExecutorService.class, defaultBindingAnnotation ) );
        scheduledExecutorService.shutdownNow();
```

**Comment by skuzzle** [comment]
Hi, thanks for the compliments. I appreciate any user and feedback that I get on my work and I'm happy to help you with this one.

You sure raise a valid point and that is certainly an evil hack. Have you tried not using the default executors at all but to create your own executor instance? Are the default executors still blocking the shutdown then if not used?

I'll explore how to best offer the shutdown option. These are my ideas:
1. A static method as you suggested is possible, but would need the `Injector` instance to get passed in
2. I could also offer a 'clean' way to inject the default executors into client code. But I feel that executor shutdown should be something that is a little more managed by the framework
3. I'm not sure whether a shutdown hook would be a good place for automatically shutting down the executors without any user interaction

I guess I'll go with the static method approach until I have a better opinion on the other bullets.

**Comment by skuzzle** [comment]
I have just implementad the static shutdown method. You can already try it using version `1.2.0-SNAPSHOT`. I'll do some more testing and adjust the documentation before releasing this.

**Comment by bgooren** [comment]
Hi, thank you for the quick response!

1) Yup, due to the executors being bound to the injector, but this is still a bit hackish
2) Wouldn't go down this route, as it exposes internals to the outside world
3) Wouldn't go down this route either, as this will not work in web applications

I think option 1 is the cleanest way to implement this from your set of ideas.

An alternative would be to expose a public interface (e.g. GuiceAsyncService) which has a start & stop method. The client code can inject it and is required to call start and stop.
This is also how another guice module I use on a daily basis manages this (guice-persist's PersistService).
It hides the internals and is a natural fit since we're using guice already anyway 👍

**Comment by skuzzle** [comment]
I like your alternative idea pretty much. Prevents the hackish passing of the injector into the static method and also allows for easier testing of client code. As of now, I'll move the `shutdownNow` method into the new interface.
I have updated the snapshot accordingly

**Comment by bgooren** [comment]
Great work 👍 👍

**Comment by skuzzle** [comment]
I have just released version `1.2.0` and updated the wiki with a paragraph about shutting down. It might take some time though until it is synced to maven central
